File: backend/app/evaluators/output.py
"""Reading structured data out of model outputs and test-case fields."""

import json
import re
from typing import Any

_FENCE_RE = re.compile(r"```(?:json|JSON)?\s*\n?(.*?)```", re.DOTALL)
_PATH_TOKEN_RE = re.compile(r"([^.\[\]]+)|\[(\d+)\]")


class JsonParseError(ValueError):
    pass


class PathNotFoundError(KeyError):
    pass


def parse_json_output(text: str, *, allow_code_fence: bool = False) -> Any:
    """Parse model output as JSON. Optionally unwraps a markdown code fence first."""
    candidate = text.strip()
    if allow_code_fence:
        fenced = _FENCE_RE.search(candidate)
        if fenced:
            candidate = fenced.group(1).strip()
    try:
        return json.loads(candidate)
    except json.JSONDecodeError as exc:
        raise JsonParseError(
            f"Output is not valid JSON: {exc.msg} (line {exc.lineno}, column {exc.colno})"
        ) from exc


def resolve_path(value: Any, path: str) -> Any:
    """Resolve a dotted path with optional list indices, e.g. ``items[0].name``."""
    current = value
    for match in _PATH_TOKEN_RE.finditer(path):
        key, index = match.groups()
        if key is not None:
            if not isinstance(current, dict) or key not in current:
                raise PathNotFoundError(path)
            current = current[key]
        else:
            position = int(index)
            if not isinstance(current, list) or position >= len(current):
                raise PathNotFoundError(path)
            current = current[position]
    return current


def to_text(value: Any) -> str:
    return value if isinstance(value, str) else json.dumps(value, ensure_ascii=False, indent=2)

File: frontend/src/demo/engine/evaluators/output.ts
/** Port of `app/evaluators/output.py`: reading structured data out of outputs and fields. */

import { isDict, pyStrip } from "../py";
import { dumps, JsonDecodeError, loads } from "../pyjson";

const FENCE_RE = /```(?:json|JSON)?\s*\n?([\s\S]*?)```/;
const PATH_TOKEN_RE = /([^.[\]]+)|\[(\d+)\]/g;

export class JsonParseError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "JsonParseError";
  }
}

export class PathNotFoundError extends Error {
  constructor(readonly path: string) {
    super(path);
    this.name = "PathNotFoundError";
  }
}

/** Parse model output as JSON, optionally unwrapping a markdown code fence first. */
export function parseJsonOutput(text: string, allowCodeFence = false): unknown {
  let candidate = pyStrip(text);
  if (allowCodeFence) {
    const fenced = FENCE_RE.exec(candidate);
    if (fenced) candidate = pyStrip(fenced[1]);
  }
  try {
    return loads(candidate);
  } catch (error) {
    if (!(error instanceof JsonDecodeError)) throw error;
    throw new JsonParseError(
      `Output is not valid JSON: ${error.msg} (line ${error.lineno}, column ${error.colno})`,
    );
  }
}

/** Resolve a dotted path with optional list indices, e.g. `items[0].name`. */
export function resolvePath(value: unknown, path: string): unknown {
  let current = value;
  for (const [, key, index] of path.matchAll(PATH_TOKEN_RE)) {
    if (key !== undefined) {
      if (!isDict(current) || !Object.hasOwn(current, key)) throw new PathNotFoundError(path);
      current = current[key];
    } else {
      const position = Number(index);
      if (!Array.isArray(current) || position >= current.length) throw new PathNotFoundError(path);
      current = current[position];
    }
  }
  return current;
}

export function toText(value: unknown): string {
  return typeof value === "string" ? value : dumps(value, { indent: 2, ensureAscii: false });
}
